text/0000-safecoin-alternative-implementation/0000-safecoin-alternative-implementation.md

```diff
@@ -0,0 +1,238 @@
+
+This means that for a section with Prefix length `n`, it will be responsible for `2^(32-n)` safecoin.
+
+Each section's CoinManagers will maintain a record `farmed` (of type `Coin`) of the amount of safecoin farmed at that section.  Any changes to this total will come as a result of an action having passed through Parsec in order to ensure that all managers maintain an eventually consistent record.
+
+The section's `farmed` value will never be allowed to exceed the amount of coins for which that section is responsible.  Ideally no section should ever get close to "running out" of farmable coins; getting the farming rate algorithm correct should ensure that.  However, in the case that a section has farmed all of its coins, it will stop issuing any more until the `farmed` value reduces again.
+
+The actual algorithm defining the farming rate is somewhat orthogonal to this RFC and worthy of its own RFC, so the details will be omitted, but it would seem fairly critical that it ensures sections don't become fully farmed or fully unfarmed.
+
+One benefit of providing coin divisibility is that on a successful farming attempt, it is trivial to pay all farmers in that section rather than the lottery approach taken in the alternative proposal whereby only one farmer is paid at a time, and only if the chosen coin is unfarmed.  Indeed, they could be paid proportional to the value they have provided to the network (this could be measured initially by node age at the moment of payment).  They could also be paid smaller amounts more frequently to encourage new Vault owners to continue to run their Vaults.
+
+To send a payment, the CoinManagers will send a `Credit` message to the destination `CoinAccount`'s section (the destination address defining the particular `CoinAccount::id`):
+
+```rust
+struct Credit {
+    amount: Coin,
+    transaction_id: Option<u64>,
+}
+```
+
+(The `transaction_id` will be explained in the "[Transferring to a Different `CoinAccount`](#transferring-to-a-different-coinaccount)" section below, but when sending farming rewards, it will always be `None`.)
+
+The CoinManager persona will be responsible for managing all aspects of farming within their section.  This will include among other things:
+
+* calculating the farming rate
+* maintaining a mapping of Vault to `CoinAccount::id` for each Vault in their section (i.e. where to pay successful farming attempts)
+* sending `Credit`s to these `CoinAccount`s when a farming attempt is successful, and increasing the `farmed` total for the section
+* receiving `Credit`s from other sections and updating the corresponding `CoinAccount`s
+
+When a section `X` splits into `X0` and `X1`, each new section will start with half of `X`'s final `farmed` value.  If `farmed.parts` is odd, rather than rounding, `X1` will be allocated the extra 1 part.
```

Comment:
Does this need to iterate over all CoinAccounts and work out the balance for each new section? What if the `X0` accounts have 55% of the farmed coins in the section? Halving the `farmed` total wouldn't be appropriate in that case, wouldn't it need to be a 55:45 split?

Reply:
No, the `farmed` value is maintained separately to the `CoinAccount`s a section manages.  If `X` has a pre-split `farmed` value of say 100 safecoin exactly, then `X0` and `X1` will start with `farmed` values of 50 safecoin each, regardless of how many `CoinAccount`s each new section ends up managing.